examples/src/read-data.c

```diff
@@ -0,0 +1,258 @@
+        }
+
+        do_pread(fd, length, offset);
+    }
+}
+
+static struct option long_opts[] = {
+    { "help", 0, 0, 'h' },
+    { "mount", 1, 0, 'm' },
+    { "offset", 1, 0, 'o' },
+    { "length", 1, 0, 'l' },
+    { 0, 0, 0, 0},
+};
+
+static char* short_opts = "hm:o:l:";
+
+static const char* usage_str =
+"\n"
+"Usage: %s [options...] <pathname>\n"
+"\n"
+"Test reading data from a file <pathname>. <pathname> should be a full\n"
+"pathname. If running without --offset and --length options, this will run\n"
+"in an interactive mode where the offset and length should be specified\n"
+"with a separating comma between them, e.g., '<offset>,<length>'.\n"
+"'quit' will terminate the program.\n"
+"\n"
+"Available options:\n"
+" -h, --help                help message\n"
+" -m, --mount=<mountpoint>  use <mountpoint> for unifyfs (default: /unifyfs)\n"
+" -o, --offset=<offset>     read from <offset>\n"
+" -l, --length=<length>     read <length> bytes\n"
+"\n";
+
+static char* program;
+
+static void print_usage(void)
+{
+    printf(usage_str, program);
+    exit(0);
+}
+
+int main(int argc, char** argv)
+{
+    int ret = 0;
+    int ch = 0;
+    int optidx = 0;
+    int unifyfs = 0;
+    int fd = -1;
+    unsigned long offset = 0;
+    unsigned long length = 0;
+    struct stat sb;
+
+    program = basename(strdup(argv[0]));
+
+    while ((ch = getopt_long(argc, argv,
+                             short_opts, long_opts, &optidx)) >= 0) {
+        switch (ch) {
+        case 'm':
+            sprintf(mountpoint, "%s", optarg);
+            break;
+
+        case 'o':
+            offset = strtoul(optarg, NULL, 0);
+            break;
+
+        case 'l':
+            length = strtoul(optarg, NULL, 0);
+            break;
+
+        case 'h':
+        default:
+            print_usage();
+            break;
+        }
+    }
+
+    if (argc - optind != 1) {
+        print_usage();
+        return -1;
+    }
+
+    sprintf(filename, "%s", argv[optind]);
+
+    if (mountpoint[0] == '\0') {
+        sprintf(mountpoint, "%s", DEFAULT_MOUNTPOINT);
+    }
+
+    if (strncmp(filename, mountpoint, strlen(mountpoint)) == 0) {
+        printf("mounting unifyfs at %s ..\n", mountpoint);
+
+        ret = unifyfs_mount(mountpoint, 0, 1, 0);
+        if (ret) {
+            fprintf(stderr, "unifyfs_mount failed (return = %d)\n", ret);
+            return -1;
+        }
+
+        unifyfs = 1;
+    }
+
+    fd = open(filename, O_RDONLY);
+    if (fd < 0) {
+        perror("open failed");
+        return -1;
+    }
+
+    ret = stat(filename, &sb);
+    if (ret < 0) {
+        perror("stat failed");
+        goto out;
+    }
+
+    printf("%s (size = %lu)\n", filename, sb.st_size);
+
+    if (offset == 0 && length == 0) {
+        run_interactive(fd);
+    } else {
+        do_pread(fd, length, offset);
+    }
+
+    ret = 0;
+out:
+    close(fd);
+
+    if (buf) {
+        free(buf);
+    }
+
+    if (unifyfs) {
+        unifyfs_unmount();
+    }
+
```

Comment:
Add:
`free(program)`

Reply:
@tonyhutter Thanks. I have updated. `man 3 basename` says not to pass the pointer to `free(3)`. So, instead, I used another pointer to keep the string returned from `strdup` to free it later.

Reply:
:+1: